### ISSUE TITLE
Leaf Profile to Interface Selector Profiles relation

### ISSUE THREAD

**Issue by Stienar** [CiscoDevNet/terraform-provider-aci#738]
It looks like "relation_infra_rs_acc_port_p" under "aci_leaf_profile" fails to create the proper relation.
If I added the relation manually from the UI next terraform apply will no longer show the bellow delta.

Terraform will perform the following actions each time i perfoam a terraform apply

// in main.tf.txt

  # aci_leaf_profile.Leaf_201 will be updated in-place
  ~ resource "aci_leaf_profile" "Leaf_201" {
        id                           = "uni/infra/nprof-Leaf_201"
        name                         = "Leaf_201"
      ~ relation_infra_rs_acc_port_p = [
          + "uni/infra/accportprof-ESXi-Hosts-vSwitch",
        ]
        # (6 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

  # aci_leaf_profile.Leaf_201_202 will be updated in-place
  ~ resource "aci_leaf_profile" "Leaf_201_202" {
        id                           = "uni/infra/nprof-Leaf_201_202"
        name                         = "Leaf_201_202"
      ~ relation_infra_rs_acc_port_p = [
          + "uni/infra/accportprof-N5k",
        ]
        # (6 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

// in vmm.tf.txt

  # aci_leaf_profile.Leaf_203_204 will be updated in-place
  ~ resource "aci_leaf_profile" "Leaf_203_204" {
        id                           = "uni/infra/nprof-Leaf_203_204"
        name                         = "Leaf_203_204"
      ~ relation_infra_rs_acc_port_p = [
          + "uni/infra/accportprof-ESXi-Hosts-VDS",
        ]
        # (6 unchanged attributes hidden)
    }

Plan: 0 to add, 3 to change, 0 to destroy

[maint.tf.txt](https://github.com/CiscoDevNet/terraform-provider-aci/files/7554929/maint.tf.txt)
[vmm.tf.txt](https://github.com/CiscoDevNet/terraform-provider-aci/files/7554930/vmm.tf.txt)

**Comment by FrederikSuijs** [comment]
Digging through Terraform Debug logs, I can find the following:
`
2021-11-17T16:45:04.755+0100 [INFO]  provider.terraform-provider-aci_v1.0.1: 2021/11/17 16:45:04 HTTP request POST /api/node/mo.json &{POST https://192.168.227.186/api/node/mo.json HTTP/1.1 1 1 map[] {{"infraRsAccPortP":{"attributes":{"annotation":"orchestrator:terraform","dn":"uni/infra/accportprof-dummy_IntProf","tDn":"uni/infra/nprof-dummy_LeafProf/rsaccPortP-[uni/infra/accportprof-dummy_IntProf]"}}}} 0x13f6700 205 [] false 192.168.227.186 map[] map[] <nil> map[]   <nil> <nil> <nil> 0xc000036080}: timestamp=2021-11-17T16:45:04.755+0100
...
2021-11-17T16:45:04.775+0100 [INFO]  provider.terraform-provider-aci_v1.0.1: 2021/11/17 16:45:04 
 HTTP response unique string POST https://192.168.227.186/api/node/mo.json {"totalCount":"1","imdata":[{"error":{"attributes":{"code":"400","text":"Invalid request. Can not contain child 'rsaccPortP-[uni\/infra\/nprof-dummy_LeafProf\/rsaccPortP-[uni\/infra\/accportprof-dummy_IntProf]]' under parent ''"}}}]}: timestamp=2021-11-17T16:45:04.775+0100
`
I'm suspecting a bug in the provider. Looks like the dn parameter in the POST is incorrect. It should be
`uni/infra/nprof-dummy_LeafProf/rsaccPortP-[uni/infra/accportprof-dummy_IntProf]`

**Comment by FrederikSuijs** [comment]
As a temporary fix/bypass, you can do the following
```terraform
resource "aci_leaf_profile" "leaf_profile" {
  name = "dummy_LeafProf"
  lifecycle {
    ignore_changes = [
      relation_infra_rs_acc_port_p,
    ]
  }
  #relation_infra_rs_acc_port_p = [
  #  aci_leaf_interface_profile.leaf_intprof.id,
  #]
}

resource "aci_rest" "leaf_profile_to_int_association" {
  path       = "/api/mo/${aci_leaf_profile.leaf_profile.id}/rsaccPortP-[${aci_leaf_interface_profile.leaf_intprof.id}].json"
  class_name = "infraRsAccPortP"
  content = {
    "annotation" : "orchestrator:terraform",
    "tDn": aci_leaf_interface_profile.leaf_intprof.id
  }
}

resource "aci_leaf_interface_profile" "leaf_intprof" {
  name = "dummy_IntProf"
}
```

**Comment by oldguardmd** [comment]
Temporary fix helped me get around this issue. Appreciate that. I lost hours wondering what I was doing wrong.

**Comment by M7Fa** [comment]
Thank you for this.  It too has been holding me up for hours trying to figure out where I went wrong.

**Comment by shrsr** [comment]
This issue has been fixed on SDK v1.20.1
The fix is incorporated in #762 and will be part of next release of the provider.

**Comment by MyUsernameIsTakenAgain** [comment]
the issue is indeed fixed by the api, i do terraform apply after I add the workaround code and the problem temporarily is solved. However for not so long, once I do terraform apply again then the interface profile just disappears 

here is the IntProf-leaf102 that was added, not it is deleted 


  # module.access-policies.aci_leaf_profile.SwProf-leaf102 will be updated in-place
  ~ resource "aci_leaf_profile" "SwProf-leaf102" {
        id                           = "uni/infra/nprof-SwProf-leaf102"
        name                         = "SwProf-leaf102"
      ~ relation_infra_rs_acc_port_p = [
          - "uni/infra/accportprof-IntProf-leaf102",
        ]
        # (5 unchanged attributes hidden)

        # (1 unchanged block hidden)

any idea ?

regards

**Comment by shrsr** [comment]
@MyUsernameIsTakenAgain Can you share your Terraform configuration file with us?

**Comment by MyUsernameIsTakenAgain** [comment]
Dear,
this is a configuration file for a switch-profile, 

resource "aci_leaf_profile" "SwProf-leaf101" {
  name        = "SwProf-leaf101"
  description  = "From Terraform"

#leaf selector under leaf profile 
  leaf_selector {
    name                    = "SwSel101"
    switch_association_type = "range"
    node_block {
      name  = "blk1"
      from_ = "101"
      to_   = "101"
    }
  
 #this one seems to have a BUG!!!, i cannot associate the interface profile inside the switch profile, if you can make the cahnges through an API instead
  
  #from interface-profile
  #relation_infra_rs_acc_card_p = [aci_leaf_interface_profile.IntProf-leaf101.id]
  }


this code I added as a workaround but did n't help as I said:


  lifecycle {
  ignore_changes = [
    relation_infra_rs_acc_card_p,
    ]
  }

resource "aci_rest" "IntProf-leaf101-via-API" {
  path       = "/api/mo/${aci_leaf_profile.SwProf-leaf101.id}/rsaccPortP-[${aci_leaf_interface_profile.IntProf-leaf101.id}].json"
  class_name = "infraRsAccPortP"
  content    = {
      "annotation" : "orchestrator:terraform",
      "tDn"        : aci_leaf_interface_profile.IntProf-leaf101.id
  }
}

this is the interface profile 

#leaf101
resource "aci_leaf_interface_profile" "IntProf-leaf101" {
    description = "From Terraform"
    name        = "IntProf-leaf101"
    #annotation  = "tag_leaf"
    #name_alias  = "name_alias"
}


Regards

**Comment by MyUsernameIsTakenAgain** [comment]
My issue is in the 
  lifecycle {
    ignore_changes = [relation_infra_rs_acc_card_p]
    }
   as it never ignore the change... the api has no issue itself.
   
   regards

**Comment by shrsr** [comment]
@MyUsernameIsTakenAgain 
You don't need to use aci_rest to create the relationship anymore. Can you please try the following configuration and let me know how it goes?

```
resource "aci_leaf_interface_profile" "IntProf-leaf101" {
  description = "From Terraform"
  name = "IntProf-leaf101"
}

resource "aci_leaf_profile" "example" {
  name = "SwProf-leaf101"
  description = "From Terraform"
  leaf_selector {
    name                    = "SwSel101"
    switch_association_type = "range"
    node_block {
      name  = "blk1"
      from_ = "101"
      to_   = "101"
    }
  }
  relation_infra_rs_acc_port_p = [
    aci_leaf_interface_profile.IntProf-leaf101.id
  ]
}
```

**Comment by MyUsernameIsTakenAgain** [comment]
Dear,
I wish that the case is as simple as you describe it,

The code, as is, can be run without any issues, see below. However the configuration never take place in the aci. so when I do terrafrom apply the first time the terrafrom will try to configure the resource, then when i do terrafrom apply again, terraform will try again to configure the resource that are not in place, and I end up in this loop, the code is trying to configure but the configuration will never happen.

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # module.access-policies.aci_leaf_profile.SwProf-leaf101 will be updated in-place
  ~ resource "aci_leaf_profile" "SwProf-leaf101" {
        id                           = "uni/infra/nprof-SwProf-leaf101"
        name                         = "SwProf-leaf101"
      ~ relation_infra_rs_acc_card_p = [
          + "uni/infra/accportprof-IntProf-leaf101",
        ]
        # (5 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.access-policies.aci_leaf_profile.SwProf-leaf101: Modifying... [id=uni/infra/nprof-SwProf-leaf101]
module.access-policies.aci_leaf_profile.SwProf-leaf101: Modifications complete after 0s [id=uni/infra/nprof-SwProf-leaf101]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

**Comment by shrsr** [comment]
@MyUsernameIsTakenAgain 
It looks like you are using a module access-policies to create resources. Can you please tell me what version of the Terraform ACI provider is being used in your main.tf of the said module?

**Comment by MyUsernameIsTakenAgain** [comment]
True,  my code is located in folder/module access-policies, and I am using the latest version of aci v2..7.0
One thing to mention, don't know whether it is relevant, I am using an ACI simulator and not a real envirnment.

Regards

**Comment by MyUsernameIsTakenAgain** [comment]
OK, but as you see that, Terrafrom doesn't complain about the code, it just carry on doing the changes yet still the resource are never configured in the aci.
Anyway, i have the resource configured under
access-policies/switch-profile.tf


resource "aci_leaf_profile" "SwProf-leaf101" {
  name        = "SwProf-leaf101"
  description  = "From Terraform"


  leaf_selector {
    name                    = "SwSel101"
    switch_association_type = "range"
    node_block {
      name  = "blk1"
      from_ = "101"
      to_   = "101"
    }
  }
  relation_infra_rs_acc_card_p = [aci_leaf_interface_profile.IntProf-leaf101.id]
}


then i have this following configuration under
access-policies/interface-profile.tf


resource "aci_leaf_interface_profile" "IntProf-leaf101" {
    description = "From Terraform"
    name        = "IntProf-leaf101"
    #annotation  = "tag_leaf"
    #name_alias  = "name_alias"
}
there should n't be any other information that is needed.

Regards

**Comment by MyUsernameIsTakenAgain** [comment]
what other information that you need ?
there is a fact that this resource had a bug and people had to develop the aci_rest to create the relationship, in somehow the issue is still there for me.

Regards

**Comment by shrsr** [comment]
![Screenshot 2023-05-23 at 8 38 53 AM](https://github.com/CiscoDevNet/terraform-provider-aci/assets/23535091/5049f8dd-4fb6-41cd-b5e8-188db9fb8f37)
![Screenshot 2023-05-23 at 8 30 35 AM](https://github.com/CiscoDevNet/terraform-provider-aci/assets/23535091/36022bc6-cb83-413b-bd50-a45fce97a507)
@MyUsernameIsTakenAgain
I tried the same config at my end and it works fine. The issue was fixed as part of #762. Please see the attached images

**Comment by MyUsernameIsTakenAgain** [comment]
I found the issue, now it is resolved, many thanks for your initiatives and time.
I was using this realtion
relation_infra_rs_acc_card_p
it suppose to be 
relation_infra_rs_acc_port_p

port ....not card....yeah yeah !
